text/0032-react-primary-framework.md

```diff
@@ -0,0 +1,105 @@
+# Making React the Primary Framework
+
+- Start Date: 2021/11/26
+- RFC PR: https://github.com/openmrs/openmrs-rfc-frontend/pull/32
+
+## Decision, including impact on distributions
+
+The app shell should already be developed using React, making React a first-class citizen. As such unnecessary framework-independent packages such as `single-spa` can be removed. Instead, `react-router` will be the central router.
+
+Existing distributions / frontend modules should not be impacted - a compatibility layer will be introduced.
+
+## Reason for decision
+
+Right now every frontend module registers a couple of components (mostly "extensions" and "pages"). Each component will be rendered using a completely new rendering tree. This comes with some significant overhead as compared to reusing a single rendering tree.
+
+Since most of our components (>= 95%) are React components and since in React its very easy to have a single rendering tree dispatching to multiple DOM nodes (using the `createPortal` API of React) we should consider going to a single rendering tree of React. The other advantage is that the framework agnostic approach has problems with otherwise simple things such as prop updates. While this is fully integrated and highly optimized in React, the framework agnostic approach right now makes the update inefficient.
+
+Besides the performance part we would also see significant reduction in complexity. Right now developers need to think in multiple dimensions. For instance, the routing is happening on an *outside* level (using `single-spa`s router) and potentially even on an *inside* level (using, e.g., the `react-router`). While most of the time this works as expected, sometimes weird behavior is observed.
+
+## Alternatives
+
+There are some ways we can go alternatively, including:
+
+1. As-is, but with enhancements to the `single-spa-react` connector (would have several downsides and would not provide all the optimizations we look for)
+2. Restricting frontend modules to React exclusively, so any module with components in another language needs either to convert or custom mount their framework / application
+
+The first way would be half-baked again, and comes with some drawbacks (frontend modules would still require recompilation here). The second way would be too much effort on the development side of frontend modules. Ideally, the current development model should not be interrupted in any way / significantly by this proposal.
+
+## Implementation guide
+
+In its core the implementation is straight forward. We create one React render tree in the `#app` element of the DOM. We'll have one global OpenMRS context just above everything with the `BrowserRouter` inside. The routes are determined by the pages inside the OpenMRS context, which enter the application through the standard registration. Therefore, instead of registering a page as an application in `single-spa`, these pages are registered in the OpenMRS context.
```

Comment:
In the described way not. Maybe a recompilation is necessary but the whole proposal was made with backwards compatibility in mind (so it would just work).

Reply:
(Just to be clear: Changes should be applied to get the most out of it. But without any changes it will still work.)